test/e2e/operator_expectations.go

```diff
@@ -173,3 +174,101 @@ func (tc *testConfig) ExpectReconcileControllersDeployment() error {
 	})
 	return err
 }
+
+func (tc *testConfig) ExpectNewNodeWhenDeletingMachine() error {
+	listOptions := client.ListOptions{
+		Namespace: namespace,
+	}
+	machineList := capiv1alpha1.MachineList{}
+	nodeList := corev1.NodeList{}
+
+	glog.Info("Get machineList")
+	err := wait.PollImmediate(1*time.Second, waitShort, func() (bool, error) {
+		if err := tc.client.List(context.TODO(), &listOptions, &machineList); err != nil {
+			glog.Errorf("error querying api for machineList object: %v, retrying...", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	glog.Info("Get nodeList")
+	err = wait.PollImmediate(1*time.Second, waitShort, func() (bool, error) {
+		if err := tc.client.List(context.TODO(), &listOptions, &nodeList); err != nil {
+			glog.Errorf("error querying api for nodeList object: %v, retrying...", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	clusterInitialTotalNodes := len(nodeList.Items)
+	clusterInitialTotalMachines := len(machineList.Items)
+	var triagedWorkerMachine capiv1alpha1.Machine
+	var triagedWorkerNode corev1.Node
+MachineLoop:
+	for _, m := range machineList.Items {
+		if m.Labels["sigs.k8s.io/cluster-api-machine-role"] == "worker" {
+			for _, n := range nodeList.Items {
+				if n.Name == m.Status.NodeRef.Name {
```

Comment:
In the tests is `NodeRef *corev1.ObjectReference `json:"nodeRef,omitempty"`` always guaranteed to be non-nil?

Reply:
If it's not, It means node Link is not working as expected so it should break

Reply:
Doesn't this mean we can (or will) deref a nil pointer then?